politeiad/backendv2/tstorebe/plugins/ticketvote/ticketvote.go

```diff
@@ -212,31 +214,220 @@ func (p *ticketVotePlugin) Hook(h plugins.HookT, payload string) error {
+		if err != nil {
+			return err
+		}
+
+		// Create inventory entry for each record.
+		ie := &invEntry{
+			data: entry{
+				Token:     hex.EncodeToString(t),
+				Status:    s.Status,
+				EndHeight: s.EndBlockHeight,
+			},
+			timestamp: r.RecordMetadata.Timestamp,
+		}
+
+		// Group inventory entries by their vote status.
+		switch {
+		case s.Status == ticketvote.VoteStatusUnauthorized:
+			unauthorized = append(unauthorized, ie)
+		case s.Status == ticketvote.VoteStatusAuthorized:
+			authorized = append(authorized, ie)
+		case s.Status == ticketvote.VoteStatusStarted:
+			started = append(started, ie)
+		case s.Status == ticketvote.VoteStatusFinished:
+			finished = append(finished, ie)
+		case s.Status == ticketvote.VoteStatusApproved:
+			approved = append(approved, ie)
+		case s.Status == ticketvote.VoteStatusRejected:
+			rejected = append(rejected, ie)
+		case s.Status == ticketvote.VoteStatusIneligible:
+			ineligible = append(ineligible, ie)
+		default:
+			return fmt.Errorf("invalid vote status for record %v",
+				ie.data.Token)
+		}
+
+		// Audit finished votes. This verifies that all cast votes use eligible
+		// tickets, and that no duplicate votes exist.
+
+		// Get vote details for eligible tickets.
+		vd, err := p.voteDetails(t)
+		if err != nil {
+			return err
+		}
+
+		// Get vote results for all cast vote details.
+		vr, err := p.voteResults(t)
+		if err != nil {
+			return err
+		}
+
+		// Create map access for the eligible tickets.
+		eligibles := make(map[string]struct{}, len(vd.EligibleTickets))
+		for _, t := range vd.EligibleTickets {
+			eligibles[t] = struct{}{}
+		}
+
+		// Range through all cast votes and make sure it was cast by a eligible
+		// ticket.
+		for _, vote := range vr {
+			_, ok := eligibles[vote.Ticket]
+			if !ok {
+				return fmt.Errorf("vote was cast by a not eligible ticket %v"+
+					"on record %v", vote.Ticket, vote.Token)
+			}
+		}
+	}
+
+	log.Infof("%v records summaries cache verified", len(tokens))
+	log.Infof("%v records audited for eligible cast votes", len(tokens))
+
+	// Verify the coherency of the submissions cache.
+	for parentToken, submissions := range rfps {
+		bToken, err := hex.DecodeString(parentToken)
+		if err != nil {
+			return err
+		}
+		cache, err := p.submissionsCache(bToken)
+		if err != nil {
+			return err
+		}
+		// Check if every submission is contained in the cache.
+		bad := false
+		for _, s := range submissions {
+			_, ok := cache.Tokens[s]
+			if ok {
+				continue
+			}
+			bad = true
+		}
+		// Check if cache is bad and needs a rebuild.
+		if bad {
+			err := p.submissionsCacheRemove(bToken)
+			if err != nil {
+				return err
+			}
+			for _, s := range submissions {
+				err = p.submissionsCacheAdd(parentToken, s)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	log.Infof("%v RFP parents submissions cache verified", len(rfps))
+
+	// Rebuild the ticketvote inventory cache.
+
+	// Sort each vote status group from oldest to newest.
```

Comment:
These are suppose to be ordered by the timestamp of their most recent vote status change.  To do that, you would order them by:

unauthorized - record timestamp 
authorized - vote authorization metadata timestamp
started - start block height
finished - end block height
approved - end block height
rejected - end block height
ineligible - record timestamp